### PR TITLE
[Feat] Multi-lora layer, data packing, optimizer, broadcast and scheduler

### PR DESCRIPTION
**Pull request by Jackmin801**
The PR currently does not do multi checkpointing and thus, checkpointing only works when the max-concurrent-runs is set to 1.

There are 3 main components to getting multi-lora to work
1. Packing and creating the right lora_cu_offsets from train idx
2. Having a MultiLoRALinear that reads off a global offset to know which LoRA to route to at any forward
3. Multi-optim and multi-broadcast. The runs object is in charge of managing per run named parameters and state dicts. It does this by having MultiLoRAModules register themselves with the global runs object after they are applied to the model. Optimizers are created using hooks that trigger when a new run is added. And schedulers are created using hooks into the optimizer creation. Broadcast has each rank get the state dict for runs that need broadcasting in unison, gathering the full tensors on master and saving to the right run.  

Few issues that motivate the choices with Runs:
1. The modules are registered instead of parameters because they do not get mutated by fsdp, compile or AC.
2. The runs syncing needs to be split into two phases, one phase where the master updates its view of active runs and another where master broadcasts the active runs to other ranks and the hook are triggered together. This is important because of the parameter reset and optimizer creation hooks acting on DTensors. Weird hanging behavior occurs when master triggers hooks and moves on before other ranks reach the hook.

The multi run test spawns an inference server and long running training server with max-concurrent-runs 2. Threee runs -- alpha, beta, gamma are created. Alpha and beta will get stepped while gamma will stall at step 2 because the trainer only accepts max-concurrent-runs 2. When alpha reaches step 10, we stop alpha and delete it's run directory, giving the slot to gamma to fill. The reward check will check that the final steps have > 0.65 reward but also that step 4 has reward between 0.2 and 0.5. This will catch when the lora is not being reset properly as step 4 will then have a high reward.

---

> [!NOTE]
> Adds end-to-end multi-run LoRA training support across packing, model, optimization, scheduling, broadcasting, and orchestration.
> 
> - Introduces `MultiLoRALinear`/`MultiLoRAModule` with global offset routing and registration to `Runs`; removes old single `LoRALinear`
> - Enhances `Runs`: distributed sync across ranks, module registration, per-run parameter access/state dicts, progress tracking, and creation hooks
> - Updates packing/data/transport: `prepare_batch` now accepts `idxs`/`num_loras`; `MicroBatch` and tensors carry `lora_num_tokens`; trainer computes per-microbatch cumulative offsets (CP-aware)
> - Adds `MultiLoRAOptimizer` and `MultiLoRAScheduler` for per-run opt/sched; `setup_optimizer`/`setup_scheduler` handle single vs multi-run
> - Filesystem broadcast now saves per-run adapter-only checkpoints via `Runs.get_state_dict_for_run` and cleans deleted run dirs
> - Adjusts perf counters for multi-LoRA and sets `RLTrainerConfig.max_concurrent_runs` default to `1`
> - CI/docs: new `Runs` docs and nav; add multi-run RL configs; add `test_rl_multi_run_lora.py` and matrixed GPU CI job
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit d7b537ff70924db90e188601d5c580259bd99a20. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: src/prime_rl/trainer/optim.py [hunk 1]
```diff
@@ -1,24 +1,94 @@
...
+        for idx in self.runs.used_idxs:
+            try:
+                self.optimizers[idx].zero_grad()
+            except Exception as e:
+                self.logger.error(f"Error zeroing grad for run {idx}: {e}")
+
+    def state_dict(self):
+        return {
+            "optimizers": [optimizer.state_dict() for optimizer in self.optimizers],
+        }
```
[the remark this answers is not in the document]
Comment (Jackmin801) [reply src/prime_rl/trainer/optim.py]: Ideally, the state dict functions are idx dependent because we want idx specific checkpoints. However, rn they arent so we need this to survive a ckpt save.

File: src/prime_rl/trainer/models/layers/lora/linear.py [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+        """Initialize LoRA parameters following standard LoRA initialization."""
+        nn.init.kaiming_uniform_(self.lora_A, a=math.sqrt(5), generator=generator)
+        nn.init.zeros_(self.lora_B)
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        """Forward pass: base_output + lora_output"""
+        base_output = self.base_layer(x)
+        lora_x = self.lora_dropout(x)
+        lora_output = (lora_x @ self.lora_A.T) @ self.lora_B.T * self.scaling
+        return base_output + lora_output
```
[the remark this answers is not in the document]
Comment (Jackmin801) [reply src/prime_rl/trainer/models/layers/lora/linear.py]: we dont plan to support this and wish to fold this into MultiLoRALinear. Every LoRAModule shoudl be multi

File: src/prime_rl/trainer/runs.py [hunk 3]
```diff
@@ -76,27 +90,53 @@ def get_orchestrator_config(self, run_id: str) -> Optional["OrchestratorConfig"]
...
+    def _delete_run_data(self, deleted_run: str, deleted_idx: int) -> None:
+        """Update data structures for a deleted run"""
+        del self.progress[deleted_idx]
+        if deleted_idx in self.config:
+            del self.config[deleted_idx]
+
+        # Process mappings
+        self.unused_idxs.add(deleted_idx)
+        del self.idx_2_id[deleted_idx]
+        del self.id_2_idx[deleted_run]
```
[the remark this answers is not in the document]
Comment (Jackmin801) [reply src/prime_rl/trainer/runs.py:102]: I need to document the ready to update cycle to be more sure that its not necessary. But atm I think it isnt necessary because 
1. We always call check for update and sync_runs sequentially
2. Before entering the calls in (1), we always have all false. In the case where the run still exists, we have completed the update and set it to false. In the case where the run dropped we errored on the update and set it to false.

File: src/prime_rl/trainer/rl/data.py [hunk 4]
```diff
@@ -35,6 +37,7 @@ def micro_batch_to_tensor(micro_batch: MicroBatch) -> TensorMicroBatch:
         inference_logprobs=torch.tensor(micro_batch.inference_logprobs, dtype=torch.float).unsqueeze(0),
         loss_mask=torch.tensor(micro_batch.loss_mask, dtype=torch.bool).unsqueeze(0),
         temperature=micro_batch.temperature if micro_batch.temperature is not None else 1.0,
+        lora_num_tokens=torch.tensor(micro_batch.lora_num_tokens, dtype=torch.int32),
```
[the remark this answers is not in the document]
Comment (Jackmin801) [reply src/prime_rl/trainer/rl/data.py]: The idea was to have a default here. I should add a default

File: src/prime_rl/trainer/runs.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (Jackmin801) [reply src/prime_rl/trainer/runs.py:102]: Added an assert before pack which should guard this

File: src/prime_rl/trainer/lora.py [hunk 5]
```diff
@@ -174,8 +178,8 @@ def apply_lora_to_model(model: nn.Module, config: LoRAConfig) -> None:
     lora_adapter_params = 0
     lora_adapted_params = 0
     for name, module in model.named_modules():
-        if isinstance(module, LoRALinear):
-            lora_adapter_params += module.lora_A.numel() + module.lora_B.numel()
+        if isinstance(module, LoRAModule):
+            lora_adapter_params += module.lora_A[0].numel() + module.lora_B[0].numel()
```
[the remark this answers is not in the document]
Comment (Jackmin801) [reply src/prime_rl/trainer/lora.py:182]: the MFU calc should be based on one tenant. not all

File: src/prime_rl/trainer/scheduler.py [hunk 6]
```diff
@@ -103,3 +105,68 @@ def setup_scheduler(
...
+            self.scheduler_config,
+            self.max_steps,
+            self.lr,
+        )
+
+    def step(self) -> None:
+        """Step all active schedulers."""
+        for idx in self.runs.used_idxs:
+            if self.schedulers[idx] is not None:
+                self.schedulers[idx].step()
```
[the remark this answers is not in the document]
Comment (Jackmin801) [reply src/prime_rl/trainer/scheduler.py]: fixed

File: src/prime_rl/trainer/rl/train.py [hunk 7]
```diff
@@ -133,7 +146,7 @@ def train(config: RLTrainerConfig):
     # Optionally, resume training from a checkpoint
     progress = Progress()
     if checkpoint_step is not None:
-        ckpt_manager.load(checkpoint_step, model, [optimizer], scheduler, progress)
+        ckpt_manager.load(checkpoint_step, model, [optimizer.optimizers[0]], scheduler, progress)
```
[the remark this answers is not in the document]
Comment (Jackmin801) [reply src/prime_rl/trainer/rl/train.py]: We now wait for hook creation in a while true loop. it aint pretty but it works.

File: src/prime_rl/trainer/batch.py [hunk 8]
```diff
@@ -57,8 +59,11 @@ def packed_samples_into_micro_bs(samples: list[MicroBatch], max_seq_len: int) ->
                 bin_content.advantages.extend(sample.advantages)
                 bin_content.inference_logprobs.extend(sample.inference_logprobs)
                 bin_content.position_ids.extend(sample.position_ids)
+                bin_content.lora_num_tokens[idx] += len(sample.input_ids)
                 break
         else:
+            sample.lora_num_tokens = [0] * num_loras
+            sample.lora_num_tokens[idx] = len(sample.input_ids)
```
[the remark this answers is not in the document]
Comment (Jackmin801) [reply src/prime_rl/trainer/batch.py:66]: Actually good catch

File: src/prime_rl/trainer/rl/train.py [hunk 9]
```diff
@@ -89,6 +88,12 @@ def train(config: RLTrainerConfig):
     )
     torch.set_float32_matmul_precision("high")
 
+    # Setup runs and offsets
+    setup_runs(config.output_dir, config.max_concurrent_runs)
+    runs = get_runs()
+    set_offsets(
+        torch.tensor([0] * config.max_concurrent_runs, dtype=torch.int32, device=torch.device("cuda", world.local_rank))
+    )
```
Comment (samsja) [inline src/prime_rl/trainer/rl/train.py]: would rename this function call, nots clear what it is doing offset is too generic

File: src/prime_rl/trainer/rl/train.py [hunk 10]
```diff
@@ -116,10 +121,23 @@ def train(config: RLTrainerConfig):
     logger.info(f"Initializing optimizer ({config.optim})")
     logger.info(f"Using `{config.loss.ratio_type}` importance ratio ({config.loss})")
 
-    optimizer = setup_optimizer(config.optim, model, parallel_dims.world_mesh["dp_shard_cp"])
+    optimizer = setup_multi_optimizer(
```
Comment (samsja) [inline src/prime_rl/trainer/rl/train.py]: if we don't have mutli run we should not be instantiating multi optimizer imo. We should probably have a config to enable multi run in which case we trigger the multi optimizer, same for multi scheduler. 

And I would rather call them multi_lora_optimizer to make it clear

File: src/prime_rl/trainer/rl/train.py [hunk 11]
```diff
@@ -116,10 +121,23 @@ def train(config: RLTrainerConfig):
...
+        # Set up the learning rate scheduler
+        while optimizer.optimizers[0] is None:
+            runs.check_for_changes()
+            runs.sync_runs()
+            logger.info(f"Waiting for optimizer to be created {runs.id_2_idx=}")
+            time.sleep(1)
+        scheduler = setup_scheduler(optimizer.optimizers[0], config.scheduler, config.max_steps, config.optim.lr)
+    else:
+        # Set up the multi-scheduler for per-run learning rate scheduling
+        scheduler = setup_multi_scheduler(config.scheduler, config.max_steps, config.optim.lr)
```
Comment (samsja) [inline src/prime_rl/trainer/rl/train.py]: would move this logic to the setup_scheduler() 

basically if we have multiple runs the setup_scheduler return the multi scheduler otherwise it return what we were returning before

File: src/prime_rl/trainer/rl/train.py [hunk 12]
```diff
@@ -116,10 +121,23 @@ def train(config: RLTrainerConfig):
...
+        # Register callback so schedulers are created when optimizers are created
+        optimizer.register_post_creation_callback(scheduler.scheduler_creation_hook)
```
Comment (samsja) [inline src/prime_rl/trainer/rl/train.py]: can we hide this in the setup_scheduler, it already taking optimizer as param)

File: src/prime_rl/trainer/lora.py [hunk 13]
```diff
@@ -174,8 +178,8 @@ def apply_lora_to_model(model: nn.Module, config: LoRAConfig) -> None:
     lora_adapter_params = 0
     lora_adapted_params = 0
     for name, module in model.named_modules():
-        if isinstance(module, LoRALinear):
-            lora_adapter_params += module.lora_A.numel() + module.lora_B.numel()
+        if isinstance(module, MultiLoRAModule):
+            lora_adapter_params += module.lora_A[0].numel() + module.lora_B[0].numel()
             lora_adapted_params += module.base_layer.weight.numel()
```
Comment (samsja) [inline src/prime_rl/trainer/lora.py:183]: so mfu can / will change when we add more lora right ? makes sense I guess

File: src/prime_rl/trainer/rl/broadcast/filesystem.py [hunk 14]
```diff
@@ -39,39 +39,48 @@ def broadcast_weights(self, model: nn.Module, step: int, adapter_only: bool = Fa
...
+                continue
+            self.logger.debug(f"Broadcasting weights for run {idx} (ready_to_update={self.runs.ready_to_update[idx]})")
 
+            if adapter_only:
+                # For adapter-only, Runs creates state dict directly for each run
+                # All ranks must participate in DTensor gathering, but only master saves
+                state_dict = self.runs.get_state_dict_for_run(idx)
+                for key, value in state_dict.items():
+                    if isinstance(value, DTensor):
+                        value = value.full_tensor()
```
[the remark this answers is not in the document]
Comment (Jackmin801) [reply src/prime_rl/trainer/rl/broadcast/filesystem.py:59]: good catch! added sort in runs

File: src/prime_rl/trainer/lora.py (hunk 13 above)
Comment (Jackmin801) [reply src/prime_rl/trainer/lora.py:183]: it might but this change is actually making it that we are agnostic to the number of concurrent lora since each token only picks one to use

File: src/prime_rl/trainer/rl/train.py [hunk 15]
```diff
@@ -117,10 +122,30 @@ def train(config: RLTrainerConfig):
...
-    optimizer = setup_optimizer(config.optim, model, parallel_dims.world_mesh["dp_shard_cp"])
+    if config.max_concurrent_runs == 1:
+        if config.model.lora:
+            # Wait for run 0 to be created in the runs system
+            while 0 not in runs.idx_2_id:
+                if world.is_master:
+                    runs.check_for_changes()
+                runs.sync_runs()
+                logger.info(f"Waiting for run 0 to be created {runs.id_2_idx=}")
+                time.sleep(1)
```
[the remark this answers is not in the document]
Comment (Jackmin801) [reply src/prime_rl/trainer/rl/train.py]: i think we can live with not benchmarking on lora with max concurrent 1

File: src/prime_rl/trainer/optim.py [hunk 16]
```diff
@@ -1,24 +1,55 @@
...
+        runs = get_runs()
+        world = get_world()
+        logger = get_logger()
+        while 0 not in runs.idx_2_id:
+            if world.is_master:
+                runs.check_for_changes()
+            runs.sync_runs()
+            logger.info(f"Waiting for run 0 to be created {runs.id_2_idx=}")
+            time.sleep(1)
+        named_params = runs.get_named_parameters_for_run(0)
```
[the remark this answers is not in the document]
Comment (Jackmin801) [reply src/prime_rl/trainer/optim.py:33]: this is fine

File: src/prime_rl/trainer/sft/train.py [hunk 17]
```diff
@@ -115,7 +115,12 @@ def train(config: SFTTrainerConfig):
 
     # Set up the optimizer
     logger.info(f"Initializing optimizer ({config.optim})")
-    optimizer = setup_optimizer(config.optim, model, parallel_dims.world_mesh["dp_shard_cp"])
+    optimizer = setup_optimizer(
+        config.optim,
+        list(model.named_parameters()),
+        parallel_dims.world_mesh["dp_shard_cp"],
+        lora=config.model.lora is not None,
+    )
```
[the remark this answers is not in the document]
Comment (Jackmin801) [reply src/prime_rl/trainer/sft/train.py]: fixed

File: src/prime_rl/trainer/optim.py (hunk 16 above)
Comment (samsja) [inline src/prime_rl/trainer/optim.py:33]: wondering if we shouldn't use the setup_multi_optimizer() even if there is only one lora.

Basically any time we actually use runs we go one path for optimizer and when we don't then the code does change.

Minor tho doesn't really matter can also merge as is

File: docs/runs.md [hunk 18]
```diff
@@ -0,0 +1,356 @@
+# Runs
```
Comment (samsja) [inline docs/runs.md:1]: do we want to keep code here btw ? I am a bit scared it might go out of sync with time

File: src/prime_rl/trainer/optim.py (hunk 16 above)
Comment (Jackmin801) [reply src/prime_rl/trainer/optim.py:33]: the idea here is to not have to maintain a separate path with runs and without runs for lora. if you are doing lora, we always assume it uses the run system even if its just one

Comment (samsja) [reply src/prime_rl/trainer/optim.py:33]: okay that's fair

**Review by samsja — APPROVED** [review]
pre-approving, LGTM. Might just need to remove code from docs

File: docs/runs.md (hunk 18 above)
Comment (Jackmin801) [reply docs/runs.md:1]: yea makes sense. I think just a high level overview of what the Runs object is responsible for is good and if contributors want to know code specifics they ask an LLM to get the latest

File: src/prime_rl/trainer/models/layers/lora/multi_linear.py [hunk 19]
```diff
@@ -0,0 +1,166 @@
...
+    ):
+        super().__init__(base_layer)
+        if rank <= 0 or n_adapters <= 0:
+            raise ValueError("rank and n_adapters must be > 0")
+
+        # Set use_grouped_mm to False if CUDA compute capability < 9.0
+        if torch.cuda.is_available():
+            cc_major, _ = torch.cuda.get_device_capability()
+            if cc_major != 9:
+                use_grouped_mm = False
```
[the remark this answers is not in the document]
Comment (Jackmin801) [reply src/prime_rl/trainer/models/layers/lora/multi_linear.py:70]: lets fix this in a follow up. im not sure if grouped_mm is supported yet
